content/posts/cross-compile-query-json.mdx

```diff
@@ -0,0 +1,252 @@
+
+```reason
+let rec compile = (expression, json) => {
+  switch (expression) {
+  | Empty => empty
+  | Keys => keys(json)
+  | Key(key, opt) => member(key, opt, json)
+  | Index(idx) => index(idx, json)
+  | Head => head(json)
+  | Tail => tail(json)
+  | Length => length(json)
+  /* [...] */
+}
+```
+
+On the left site are defined all the possible `Variants` and on the right site the operations. Those operations transforms the JSON. Here is where `map`, `filter`, `reduce`, `index`, etc... are implemented.
+
+### Running
+The easier part, the compile step give us back a curried function that expects a json as only argument, we just apply the function to this JSON and print the result.
+
+This example only describe the happy path, in reality the parsing and compilation steps return a `Result` type which allows handling errors more composable.
+
+# Distribution
+Now we already covered how it works internally and a little overview of the architecture. Now let's dive on how developers can use it in their machines.
+
+## How to cross-compilation to binary
+query-json uses `esy` as a package manager that helps managing either native OCaml and Reason from opam registry or JavaScript packages from npm registry; and `dune` as a build system.
+
+All of the build process and their tests runs on our CI in Github Actions, which allows to run Mac, Windows and Linux images to compile with the OCaml compiler to different architectures.
+
+Once the build is success it pushes the binary into Github Releases and npm registry.
+
+## How compile to the Web
+The compilation to JavaScript is the sweet section of this blog post, since we are using OCaml (under the hood while writting Reason), it allows us to compile directlty to JavaScript using [js_of_ocaml](https://github.com/ocsigen/js_of_ocaml).
+
+Using js_of_ocaml in dune is as easy as adding a flag in the dune file. Using the same build command you have an artifact in JavaScript 🤯 . It will compile any library to JavaScript.
+
+As you can see, query-json uses menhir, sedlex and [yojson](https://github.com/ocaml-community/yojson)
+
+In order to use it in esy, I only needed to add
+```bash
+esy add js_of_ocaml
+esy add js_of_ocaml-compiler
+```
+
+```clojure
+(executable
+ (name Js)
+ (modes js)
+ (libraries console.lib source yojson js_of_ocaml)
+)
+```
+
+## Made query-json's playground
+
+After having a JavaScript artifact I was able to run it as playground to teach people how to use it or to test new versions on each pull request in the browser.
+
+I build a website using Reason and BuckleScript which uses query-json JavaScript's artifact to run it, you can try it yourself here:
+
+https://query-json.netlify.app
+
+The query-json computation runs sincronous since it's able to run on each key-stroke. Comparing this playground with [jqplay.org](https://jqplay.org/)
```

Comment:
```suggestion
The query-json computation runs synchronously since it's able to run on each key-stroke. Comparing this playground with [jqplay.org](https://jqplay.org/)
```